docs/reference/configuration.md
# Configuration

This page describes how to set different settings in immudb.

Some of the most important settings that ones are:

| Parameter           | Default   | Description                                                                                                  |
| ------------------- | --------- | ------------------------------------------------------------------------------------------------------------ |
| `dir`               | `./data`  | System and user databases are stored here                                                                    |
| `network`           | `tcp`     |                                                                                                              |
| `address`           | `0.0.0.0` | Listening address                                                                                            |
| `port`              | `3322`    | Listing port                                                                                                 |
| `mtls`              | `false`   | Whether to enable [Mutual TLS (opens new window)](https://en.wikipedia.org/wiki/Mutual\_authentication#mTLS) |
| `pkey`              |           | If specified, the server can sign the state the clients use to verify immutability                           |
| `auth`              | `true`    | If enabled, immudb will require user and password from the client                                            |
| `clientcas`         |           | Client certificate authority                                                                                 |
| `maintenance`       |           | Maintenance mode. Override the authentication flag                                                           |
| `sync`              |           | Runs in sync mode. Prevents data loss but affects performance                                                |
| `token-expiry-time` | `1440`    | Client token expiry time, in minutes                                                                         |
| `web-server`        | `true`    | Embedded web console server                                                                                  |
| `web-server-port`   | `8080`    | Embeded web console port server                                                                              |
| `pgsql-server`      | `true`    | pqsql protocol compatibility server (allows to connect from pgsql compatible clients)                        |
| `pgsql-server-port` | `5432`    | pqsql protocol compatibility server port                                                                     |

Settings can be specified as command line options to immudb (see `immudb -h`), in a configuration file, or as environment variables.

### Configuration file <a href="#configuration-file" id="configuration-file"></a>

Settings can be specified in a [immudb.toml configuration file (opens new window)](https://raw.githubusercontent.com/codenotary/immudb/master/configs/immudb.toml).

Which configuration file to use is set with the `--config` option. By default, immudb looks into the `configs` subfolder in the current directory.

When running immudb as a service, `immudb service install` allows to specify the configuration file to use with the `--config` option.

### Environment variables <a href="#environment-variables" id="environment-variables"></a>

Settings specified via environment variables take override the configuration file. They are specified in the form of `IMMUDB_`, for example `IMMUDB_DIR` specifies the `dir` variable.


CHANGELOG.md
# CHANGELOG
All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
<a name="unreleased"></a>
## [Unreleased]


<a name="v1.2.1"></a>
## [v1.2.1] - 0001-01-01

<a name="v1.2.0"></a>
## [v1.2.0] - 2021-12-13
### Bug Fixes
- fix interactive use database
- **database:** Internal consistency check on data reads
- **database/meta:** Do not crash on history with deleted items
- **embedded/store:** change already closed error message
- **embedded/store:** readonly tx entries to ensure no runtime modification
- **embedded/store:** reserve 4bytes in buffers for nentries
- **embedded/tbtree:** set fixed snapshot ts
- **pkg/database:** history skipping not found entries
- **pkg/server/sessions:** remove transaction on read conflict error
- **pkg/server/sessions/internal/transactions:** transaction is cleared after sqlExec error
- **protobuf:** Fix compatibility with 1.1 version
- **sql:** Do not panic on error during delete
- **tx:** Remove summary from metadata

### Changes
- **cmd/immuadmin/command:** add super user login hint
- **embedded/sql:** support for escaped strings
- **embedded/sql:** use sql standard escaping with single quotes
- **embedded/store:** prevent value reading of expired entries
- **embedded/store:** private readonly metadata is validated when reading data
- **embedded/store:** mandatory expiration filter
- **embedded/store:** read-only kv metadata for committed entries
- **embedded/store:** rw and readonly kv metadata
- **embedded/store:** txmetdata placeholder with zero len
- **embedded/store:** easily extendable meta attributes
- **embedded/store:** reduce attribute code size
- **embedded/store:** return data corrupted error when deserialization cannot proceed
- **embedded/store:** use fixed time during the lifespan of a tx
- **embedded/store:** validations during metadata deserialization
- **embedded/store:** improve metadata serialization/deserialization methods
- **embedded/store:** dedicated expiration error
- **embedded/store:** fix expiration error declaration
- **makefile:** remove windows binaries digital signature
- **pkg/api:** use new kvmetadata api
- **pkg/auth:** require admin permission to export and replicate txs
- **pkg/client:** tx read conflict error is mapped in an CodInFailedSqlTransaction
- **pkg/integration:** remove useless compilation tag on tests
- **pkg/server:** deprecate GetAuth and WithAuth
- **pkg/server/sessions:** tuning sessions params
- **pkg/server/sessions:** session timeout set to 2 min
- **pkg/server/sessions:** session max inactivity time set to 3m and minor stat collecting fix
- **pkg/server/sessions/internal/transactions:** defer only when needed
- **pkg/stdlib:** clean tx after rollback
- **pkg/stdlib:** fix connection creation
- **server/sessions:** modify read conflict error message

### Features
- **embedded/store:** logical entries expiration
- **pkg/api:** logical entries expiration
- **pkg/client:** expirable set
- **pkg/stdlib:** expose tx on std lib


<a name="v1.2.0-RC1"></a>
## [v1.2.0-RC1] - 2021-12-07
### Bug Fixes
- Update jaswdr/faker to v1.4.3 to fix build on 32-bit systems
- **Makefile:** Fix building immudb for specific os/arch
- **Makefile:** Use correct version of the grpc-gateway package
- **embedded/sql:** ensure determinism and no value overlaps distinct rows
- **embedded/sql:** fix inserting calculated null values
- **embedded/sql:** correct max key length validation based on specified col max length
- **embedded/sql:** fix rollback stmt
- **embedded/sql:** normalize parameters with lower case identifiers
- **embedded/sql:** param substitution in LIKE expression
- **embedded/sql:** Use correct statement for subquery
- **embedded/sql:** Do not modify value returned by colsBySelector
- **embedded/sql:** Fix SELECT * when joining with subquery
- **embedded/sql:** distinct row reader with limit argument
- **embedded/store:** release lock when tx has a conflict
- **embedded/store:** read conflict validation
- **embedded/store:** typo in error message
- **pkg/auth:** Fix password tests
- **pkg/client:** fix database name saving on token service
- **pkg/database:** sql exec on provided tx
- **pkg/server:** fix keep alive session interceptor
- **testing:** using pointers for job channels
- **webconsole:** Fix html of the default missing page.

### Changes
- refining sdk client constructor and add readOnly tx guard
- Update build/RELEASING.md documentation.
- fix more tests
- decoupled token service
- token is handled internally by sdk. Remove useless code
- remove token service from client options and fix tests
- **cmd/immuadmin/command:** fix immuadmin token name on client creation
- **cmd/immuclient:** deleteKeys functioin and updates after metadata-related changes
- **cmd/immuclient:** temporary disable displaying hash in non-verified methods
- **embeddded/tbtree:** leverage snapshot id to identify it's the current unflushed one
- **embedded/multierr:** minor code simplification
- **embedded/sql:** set INNER as default join type
- **embedded/sql:** minor update after rebasing
- **embedded/sql:** Alter index key prefixes
- **embedded/sql:** wip rw transactions
- **embedded/sql:** remove opt_unique rule to ensure proper error message
- **embedded/sql:** minor code simplification
- **embedded/sql:** kept last snapshot open
- **embedded/sql:** Simplify row_reader key selection
- **embedded/sql:** Better error messages when (up|in)serting data
- **embedded/sql:** standardized datasource aliasing
- **embedded/sql:** wip sqlTx
- **embedded/sql:** de-duplicate tx attributes using tx header struct
- **embedded/sql:** fix nullable values handling
- **embedded/sql:** rollback token
- **embedded/sql:** set parsing verbose mode when instantiating sql engine
- **embedded/sql:** unsafe snapshot without flushing
- **embedded/sql:** reusable index entries and ignore deleted index entries

pkg/server/options.go
/*
Copyright 2022 CodeNotary, Inc. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

	http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package server

import (
	"crypto/tls"
	"fmt"
	"net"
	"strconv"
	"strings"
	"time"

	"github.com/codenotary/immudb/pkg/server/sessions"

	"github.com/codenotary/immudb/pkg/stream"

	"github.com/codenotary/immudb/embedded/store"
	"github.com/codenotary/immudb/pkg/auth"
)

const SystemDBName = "systemdb"
const DefaultDBName = "defaultdb"
const DefaultMaxValueLen = 1 << 25   //32Mb
const DefaultStoreFileSize = 1 << 29 //512Mb

// Options server options list
type Options struct {
	Dir                  string
	Network              string
	Address              string
	Port                 int
	MetricsPort          int
	Config               string
	Pidfile              string
	Logfile              string
	TLSConfig            *tls.Config
	auth                 bool
	MaxRecvMsgSize       int
	NoHistograms         bool
	Detached             bool
	MetricsServer        bool
	WebServer            bool
	WebServerPort        int
	DevMode              bool
	AdminPassword        string `json:"-"`
	systemAdminDBName    string
	defaultDBName        string
	listener             net.Listener
	usingCustomListener  bool
	maintenance          bool
	SigningKey           string
	synced               bool
	RemoteStorageOptions *RemoteStorageOptions
	StreamChunkSize      int
	TokenExpiryTimeMin   int
	PgsqlServer          bool
	PgsqlServerPort      int
	ReplicationOptions   *ReplicationOptions
	SessionsOptions      *sessions.Options
}

type RemoteStorageOptions struct {
	S3Storage     bool
	S3Endpoint    string
	S3AccessKeyID string
	S3SecretKey   string `json:"-"`
	S3BucketName  string
	S3PathPrefix  string
}

type ReplicationOptions struct {
	MasterAddress    string
	MasterPort       int
	MasterDatabase   string
	FollowerUsername string
	FollowerPassword string
}

// DefaultOptions returns default server options
func DefaultOptions() *Options {
	return &Options{
		Dir:                  "./data",
		Network:              "tcp",
		Address:              "0.0.0.0",
		Port:                 3322,
		MetricsPort:          9497,
		WebServerPort:        8080,
		Config:               "configs/immudb.toml",
		Pidfile:              "",
		Logfile:              "",
		TLSConfig:            nil,
		auth:                 true,
		MaxRecvMsgSize:       1024 * 1024 * 32, // 32Mb
		NoHistograms:         false,
		Detached:             false,
		MetricsServer:        true,
		WebServer:            true,
		DevMode:              false,
		AdminPassword:        auth.SysAdminPassword,
		systemAdminDBName:    SystemDBName,
		defaultDBName:        DefaultDBName,
		usingCustomListener:  false,
		maintenance:          false,
		synced:               true,
		RemoteStorageOptions: DefaultRemoteStorageOptions(),
		StreamChunkSize:      stream.DefaultChunkSize,
		TokenExpiryTimeMin:   1440,
		PgsqlServer:          false,
		PgsqlServerPort:      5432,
		SessionsOptions:      sessions.DefaultOptions(),
	}
}

func (opts *Options) DefaultStoreOptions() *store.Options {
	indexOptions := store.DefaultIndexOptions().
		WithRenewSnapRootAfter(0).
		WithCompactionThld(0).
		WithDelayDuringCompaction(10 * time.Millisecond)

	return store.DefaultOptions().
		WithIndexOptions(indexOptions).
		WithMaxLinearProofLen(0).
		WithMaxConcurrency(30).
		WithMaxIOConcurrency(1).
		WithFileSize(DefaultStoreFileSize).
		WithMaxKeyLen(store.DefaultMaxKeyLen).
		WithMaxValueLen(DefaultMaxValueLen).
		WithMaxTxEntries(store.DefaultMaxTxEntries).
		WithSynced(opts.synced)
}

func DefaultRemoteStorageOptions() *RemoteStorageOptions {
	return &RemoteStorageOptions{
		S3Storage: false,
	}
}

// WithDir sets dir
func (o *Options) WithDir(dir string) *Options {
	o.Dir = dir
	return o
}

// WithNetwork sets network
func (o *Options) WithNetwork(network string) *Options {
	o.Network = network
	return o
}

// WithAddress sets address
func (o *Options) WithAddress(address string) *Options {
	o.Address = address
	return o
}

// WithPort sets port
func (o *Options) WithPort(port int) *Options {
	o.Port = port
	return o
}

// WithConfig sets config file name
func (o *Options) WithConfig(config string) *Options {
	o.Config = config
	return o
}

// WithPidfile sets pid file
func (o *Options) WithPidfile(pidfile string) *Options {
	o.Pidfile = pidfile
	return o
}

// WithLogfile sets logfile
func (o *Options) WithLogfile(logfile string) *Options {
	o.Logfile = logfile
	return o
}

// WithTLS sets tls config
func (o *Options) WithTLS(tls *tls.Config) *Options {
	o.TLSConfig = tls
	return o
}

// WithAuth sets auth
// Deprecated: WithAuth will be removed in future release
func (o *Options) WithAuth(authEnabled bool) *Options {
	o.auth = authEnabled
	return o
}

func (o *Options) WithMaxRecvMsgSize(maxRecvMsgSize int) *Options {
	o.MaxRecvMsgSize = maxRecvMsgSize
	return o
}

// GetAuth gets auth
// Deprecated: GetAuth will be removed in future release
func (o *Options) GetAuth() bool {
	return o.auth
}

// WithNoHistograms disables collection of histograms metrics (e.g. query durations)
func (o *Options) WithNoHistograms(noHistograms bool) *Options {
	o.NoHistograms = noHistograms
	return o
}

// WithDetached sets immudb to be run in background
func (o *Options) WithDetached(detached bool) *Options {
	o.Detached = detached
	return o
}

// Bind returns bind address
func (o *Options) Bind() string {
	return o.Address + ":" + strconv.Itoa(o.Port)
}

// MetricsBind return metrics bind address
func (o *Options) MetricsBind() string {
	return o.Address + ":" + strconv.Itoa(o.MetricsPort)
}

// WebBind return bind address for the Web API/console
func (o *Options) WebBind() string {
	return o.Address + ":" + strconv.Itoa(o.WebServerPort)
}

// String print options
func (o *Options) String() string {
	rightPad := func(k string, v interface{}) string {
		return fmt.Sprintf("%-17s: %v", k, v)
	}
	opts := make([]string, 0, 17)
	opts = append(opts, "================ Config ================")
	opts = append(opts, rightPad("Data dir", o.Dir))
	opts = append(opts, rightPad("Address", fmt.Sprintf("%s:%d", o.Address, o.Port)))

	repOpts := o.ReplicationOptions

	if o.ReplicationOptions != nil {
		opts = append(opts, rightPad("Replica of", fmt.Sprintf("%s:%d", repOpts.MasterAddress, repOpts.MasterPort)))
	}

	if o.MetricsServer {
		opts = append(opts, rightPad("Metrics address", fmt.Sprintf("%s:%d/metrics", o.Address, o.MetricsPort)))
	}
	if o.Config != "" {
		opts = append(opts, rightPad("Config file", o.Config))
	}
	if o.Pidfile != "" {
		opts = append(opts, rightPad("PID file", o.Pidfile))
	}
	if o.Logfile != "" {
		opts = append(opts, rightPad("Log file", o.Logfile))
	}
	opts = append(opts, rightPad("Max recv msg size", o.MaxRecvMsgSize))
	opts = append(opts, rightPad("Auth enabled", o.auth))
	opts = append(opts, rightPad("Dev mode", o.DevMode))
	opts = append(opts, rightPad("Default database", o.defaultDBName))
	opts = append(opts, rightPad("Maintenance mode", o.maintenance))
	opts = append(opts, rightPad("Synced mode", o.synced))
	if o.RemoteStorageOptions.S3Storage {
		opts = append(opts, "S3 storage")
		opts = append(opts, rightPad("   endpoint", o.RemoteStorageOptions.S3Endpoint))
		opts = append(opts, rightPad("   bucket name", o.RemoteStorageOptions.S3BucketName))
		opts = append(opts, rightPad("   prefix", o.RemoteStorageOptions.S3PathPrefix))
	}
	if o.AdminPassword == auth.SysAdminPassword {
		opts = append(opts, "----------------------------------------")
		opts = append(opts, "Superadmin default credentials")
		opts = append(opts, rightPad("   Username", auth.SysAdminUsername))
		opts = append(opts, rightPad("   Password", auth.SysAdminPassword))
	}
	opts = append(opts, "========================================")
	return strings.Join(opts, "\n")
}

// WithMetricsServer ...
func (o *Options) WithMetricsServer(metricsServer bool) *Options {
	o.MetricsServer = metricsServer
	return o
}

// WithWebServer ...
func (o *Options) WithWebServer(webServer bool) *Options {
	o.WebServer = webServer
	return o
}

// WithWebServerPort ...
func (o *Options) WithWebServerPort(port int) *Options {
	o.WebServerPort = port
	return o
}

// WithDevMode ...
func (o *Options) WithDevMode(devMode bool) *Options {
	o.DevMode = devMode
	return o
}

// WithAdminPassword ...
func (o *Options) WithAdminPassword(adminPassword string) *Options {
	o.AdminPassword = adminPassword
	return o
}

//GetSystemAdminDBName returns the System database name
func (o *Options) GetSystemAdminDBName() string {
	return o.systemAdminDBName
}

//GetDefaultDBName returns the default database name
func (o *Options) GetDefaultDBName() string {
	return o.defaultDBName
}

// WithListener used usually to pass a bufered listener for testing purposes
func (o *Options) WithListener(lis net.Listener) *Options {
	o.listener = lis
	o.usingCustomListener = true
	return o
}

// WithMaintenance sets maintenance mode
func (o *Options) WithMaintenance(m bool) *Options {
	o.maintenance = m
	return o
}

// GetMaintenance gets maintenance mode
func (o *Options) GetMaintenance() bool {
	return o.maintenance
}

// WithSynced sets synced mode
func (o *Options) WithSynced(synced bool) *Options {
	o.synced = synced
	return o
}

// GetSynced gets synced mode
func (o *Options) GetSynced() bool {
	return o.synced
}

// WithSigningKey sets signature private key
func (o *Options) WithSigningKey(signingKey string) *Options {
	o.SigningKey = signingKey
	return o
}

// WithStreamChunkSize set the chunk size
func (o *Options) WithStreamChunkSize(streamChunkSize int) *Options {
	o.StreamChunkSize = streamChunkSize
	return o
}

// WithTokenExpiryTime set authentication token expiration time in minutes
func (o *Options) WithTokenExpiryTime(tokenExpiryTimeMin int) *Options {
	o.TokenExpiryTimeMin = tokenExpiryTimeMin
	return o
}

// PgsqlServerPort enable or disable pgsql server
func (o *Options) WithPgsqlServer(enable bool) *Options {
	o.PgsqlServer = enable
	return o
}

// PgsqlServerPort sets pgdsql server port
func (o *Options) WithPgsqlServerPort(port int) *Options {
	o.PgsqlServerPort = port
	return o
}

func (o *Options) WithRemoteStorageOptions(remoteStorageOptions *RemoteStorageOptions) *Options {
	o.RemoteStorageOptions = remoteStorageOptions
	return o
}

func (o *Options) WithReplicationOptions(replicationOptions *ReplicationOptions) *Options {
	o.ReplicationOptions = replicationOptions
	return o
}

func (o *Options) WithSessionOptions(options *sessions.Options) *Options {
	o.SessionsOptions = options
	return o
}

// RemoteStorageOptions

func (opts *RemoteStorageOptions) WithS3Storage(S3Storage bool) *RemoteStorageOptions {
	opts.S3Storage = S3Storage
	return opts
}

func (opts *RemoteStorageOptions) WithS3Endpoint(s3Endpoint string) *RemoteStorageOptions {
	opts.S3Endpoint = s3Endpoint
	return opts
}

func (opts *RemoteStorageOptions) WithS3AccessKeyID(s3AccessKeyID string) *RemoteStorageOptions {
	opts.S3AccessKeyID = s3AccessKeyID
	return opts
}

func (opts *RemoteStorageOptions) WithS3SecretKey(s3SecretKey string) *RemoteStorageOptions {
	opts.S3SecretKey = s3SecretKey
	return opts
}

func (opts *RemoteStorageOptions) WithS3BucketName(s3BucketName string) *RemoteStorageOptions {
	opts.S3BucketName = s3BucketName
	return opts
}

func (opts *RemoteStorageOptions) WithS3PathPrefix(s3PathPrefix string) *RemoteStorageOptions {
	opts.S3PathPrefix = s3PathPrefix
	return opts
}

// ReplicationOptions

func (opts *ReplicationOptions) WithMasterAddress(masterAddress string) *ReplicationOptions {
	opts.MasterAddress = masterAddress
	return opts
}

func (opts *ReplicationOptions) WithMasterPort(masterPort int) *ReplicationOptions {
	opts.MasterPort = masterPort
	return opts
}

func (opts *ReplicationOptions) WithMasterDatabase(masterDatabase string) *ReplicationOptions {
	opts.MasterDatabase = masterDatabase
	return opts
}

func (opts *ReplicationOptions) WithFollowerUsername(followerUsername string) *ReplicationOptions {
	opts.FollowerUsername = followerUsername
	return opts
}

func (opts *ReplicationOptions) WithFollowerPassword(followerPassword string) *ReplicationOptions {
	opts.FollowerPassword = followerPassword
	return opts
}
